### ISSUE TITLE
New lint: `slice_indexing_after_deconstruction`

### ISSUE THREAD

**Issue by xFrednet** [rust-lang/rust-clippy#7569]
### What it does

The lint checks for slices that come from value deconstruction and are only used to access individual slice values.

### Categories (optional)

- Kind: *pedantic* (This could also be *style*, but I'm not sure if we want it warn-by-default.)

### Advantages

Deconstruction of slices allows direct value access without using indexes on the slice, which can lead to panics. It also enables the user to give slice values a better name and indirectly check the correct length of the slice. See [\#7566 files](https://github.com/rust-lang/rust-clippy/pull/7566/files) as an example of how this can be an improvement.

Clippy already had a few instances where ICEs would have been prevented by using deconstruction of slices. 

### Example

```rust
fn main() {
    let slice: Option<&[u32]> = Some(&[1, 2, 3]);
    
    if let Some(slice) = slice {
        println!("{}", slice[0]);
    }
}
```

Could be written as:

```rust
fn main() {
    let slice: Option<&[u32]> = Some(&[1, 2, 3]);
    
    if let Some([first, ..]) = slice {
        println!("{}", first);
    }
}
```

### Notes:

* Clippy uses pattern matching extensively, often in  `if_chain` macros with `if let` statements. I think it would be reasonable for this lint to also check if statements inside `if_chain` macros.
* The lint probably shouldn't warn if the slice instance is used for something else besides direct value access.

**Comment by camsteffen** [comment]
We have to pick some max threshold for indexes to lint. Like we don't want to lint `slice[100]`. Maybe `slice[3]` is a reasonable max to lint? In other words, suggest a slice pattern with up to 4 elements.

**Comment by xFrednet** [comment]
Good suggestion, a threshold of 3 would catch all instance in Clippy that I've seen. We could also have this configurable and have the default be 3 :+1:.
